[6.2] [nonisolated-nonsending] Make the AST not consider nonisolated(nonsending) to be an actor isolation crossing point.

gottesmm opened the pull request:
Explanation: Big picture this fixes a large hole in the implementation of
SE-0461 and then fixes a few additional issues that this exposed.

Previously Sema was allowing for nonisolated(nonsending) callees to be potential
sending points. This violated the semantics of SE-0461. We previously got away
with this since Sema was delegating checking of function arguments to
SendNonSendable and SendNonSendable was treating them as isolated functions due
to their isolated parameter and thus squelched the error.

Our luck ran out when it came to imported objc async methods. These do not
have an isolated parameter, so we would consider them to be nonisolated at the
SendNonSendable level. So we did not squelch the error. So for instance, the
following function would emit an error when it clearly should not:

```swift
@MainActor
func defeatRegion(baz: Baz) async throws -> Bar? {
  let foo = ObjectiveCType()
  return try await foo.findCompatibleBar(for: baz) // ERROR: Sending 'baz' risks causing data races
}
```

In this change, we fix that problem by treating nonisolated(nonsending) callees
as not being a send since we are going to be inheriting isolation from our
caller.

In the process of doing this, the author looked through this code to see if
there were any other issues that violated SE-0461 and found that we were
allowing for parameters of nonisolated(nonsending) parameters to be passed to
@concurrent functions which should not be allowed. The reason why this is not
allowed is that the parameter of the nonisolated(nonsending) function /could/ be
isolated to an actor meaning that by passing it to the @concurrent function we
would be escaping the value to another isolation domain. We fixed this as well
at the Sema level.

In order to be careful with this change, the authors massively expanded the test
coverage for this feature by porting all of the SendNonSendable tests to run
both with and without NonisolatedNonsendingByDefault.

When I enabled those tests, we discovered additional issues exposed by inferring nonisolated(nonsending) on expressions (#82731) . Specifically:

1. We did not handle cases where a closure would be global actor isolated or isolated to an actor instance due to a capture. I fixed this in https://github.com/swiftlang/swift/pull/82776/commits/78bc443ace3b7cc73a92a40bd24930ccddccb23b
2. We were introducing round trip thunks that went from `nonisolated(nonsending)` -> `@concurrent` -> `nonisolated(nonsending)` which results in a nonisolated(nonsending) function that always has concurrent isolation (which isn't what we want). The peephole eliminates that issues.
3. We were not respecting the isolation of inheritActorIsolation. Instead we always did nonisolated(nonsending) in such cases resulting in the incorrect isolation.

With all of those together, we now pass all of the transfernonsendable tests with NonisolatedNonsendingByDefault enabled.

When doing that, the authors noticed that when NonisolatedNonsendingByDefault
was enabled, we were emitting diagnostics saying caller inheriting
isolation. This violates the spirit of SE-0461, so we threaded through a
SILFunction into the diagnostic printing code and used that to tweak the
diagnostics depending on whether or not NonisolatedNonsendingByDefault is
enabled. In the case where it is disabled:

1. Functions with @concurrent isolation are called "nonisolated" and regions
containing parameters of those functions are called "task-isolated".
5. Functions with nonisolated(nonsending) isolation are called
"nonisolated(nonsending)" and regions containing parameters of those functions
are called "nonisolated(nonsending) task-isolated".

When it is enabled:

1. Functions with @concurrent isolation are called "@concurrent" and regions
containing parameters of those functions are called "@concurrent task-isolated".
2. Functions with nonisolated(nonsending) isolation are called
"nonisolated" and regions containing parameters of those functions
are called "task-isolated".

This ensures that in cases that do not involve explicit @concurrent or
nonisolated(nonsending) parameters, one just sees task-isolated and nonisolated
in the SendNonSendable diagnostics when one enables or disables
NonisolatedNonsendingByDefault due to the way the upcoming feature changes how
we infer isolation.

Scope: This impacts the diagnostics that we emit when calling
nonisolated(nonsending) functions and when passing a parameter of an
nonisolated(nonsending) function to an @concurrent function.

Issues:
- rdar://154139237


Original PRs:

- #82555
- https://github.com/swiftlang/swift/pull/82858

Risk: Medium. The main risk is that we would start emitting less diagnostics
that we did not perviously when calling nonisolated(nonsending) function or
calling nonisolated functions from nonisolated(nonsending) functions. That being
said this closes a few major holes in the model and ensures that we actually
follow the semantics of SE-0461 so the authors believe it is worth it. The
authors also believe that by massively expanding the amount of tests that run
with this feature that we have mitigated the risk of this fix.

Testing: Massively expanded the amount of tests that run with
NonisolatedNonsendingByDefault enabled by converting all SendNonSendable tests
(and a few others) to run with both. Also added specific tests to test out this
behavior both with and without objc.

Reviewer: @xedin

AnthonyLatsis committed:
AST: Cut down on `DescriptiveDeclKind` usage in `DiagnosticsSIL.def`

(cherry picked from commit 5c190b961300202886cfa1970b8506344853f21f)

gottesmm committed:
[nonisolated-nonsending] Make the AST not consider nonisolated(nonsending) to be an actor isolation crossing point.

We were effectively working around this previously at the SIL level. This caused
us not to obey the semantics of the actual evolution proposal. As an example of
this, in the following, x should not be considered main actor isolated:

```swift
nonisolated(nonsending) func useValue<T>(_ t: T) async {}
@MainActor func test() async {
  let x = NS()
  await useValue(x)
  print(x)
}
```

we should just consider this to be a merge and since useValue does not have any
MainActor isolated parameters, x should not be main actor isolated and we should
not emit an error here.

I also fixed a separate issue where we were allowing for parameters of
nonisolated(nonsending) functions to be passed to @concurrent functions. We
cannot allow for this to happen since the nonisolated(nonsending) parameters
/could/ be actor isolated. Of course, we have lost that static information at
this point so we cannot allow for it. Given that we have the actual dynamic
actor isolation information, we could dynamically allow for the parameters to be
passed... but that is something that is speculative and is definitely outside of
the scope of this patch.

rdar://154139237
(cherry picked from commit c12c99fb734b7d9b9ba36c58956b5aa1f74ff4ba)

gottesmm committed:
[rbi] Convert all rbi tests to run also in NonisolatedNonsendingByDefault.

Going to update the tests in the next commit. This just makes it easier to
review.

(cherry picked from commit a6edf4fb909832fce5db8e364f8d1f346dbf9615)

gottesmm committed:
[rbi] Wrap use ActorIsolation::printForDiagnostics with our own SILIsolationInfo::printActorIsolationForDiagnostics.

I am doing this so that I can change how we emit the diagnostics just for
SendNonSendable depending on if NonisolatedNonsendingByDefault is enabled
without touching the rest of the compiler.

This does not actually change any of the actual output though.

(cherry picked from commit 4ce4fc4f954da8fe95735ae9d162bd5eb34ca8c0)

gottesmm committed:
[rbi] Begin tracking if a function argument is from a nonisolated(nonsending) parameter and adjust printing as appropriate.

Specifically in terms of printing, if NonisolatedNonsendingByDefault is enabled,
we print out things as nonisolated/task-isolated and @concurrent/@concurrent
task-isolated. If said feature is disabled, we print out things as
nonisolated(nonsending)/nonisolated(nonsending) task-isolated and
nonisolated/task-isolated. This ensures in the default case, diagnostics do not
change and we always print out things to match the expected meaning of
nonisolated depending on the mode.

I also updated the tests as appropriate/added some more tests/added to the
SendNonSendable education notes information about this.

(cherry picked from commit 14634b684737a103dd20d25eaa8fc4550a4b5fe4)

gottesmm committed:
[rbi] Thread through a SILFunction into print routines so we can access LangOpts.Features so we can change how we print based off of NonisolatedNonsendingByDefault.

We do not actually use this information yet though... This is just to ease
review.

(cherry picked from commit 4433ab8d81c3ea3e76320e4bd8c63f09b4db2954)

gottesmm committed:
[rbi] Use interned StringRefs for diagnostics instead of SmallString<64>.

This makes the code easier to write and also prevents any lifetime issues from a
diagnostic outliving the SmallString due to diagnostic transactions.

(cherry picked from commit 010fa39f31e051a33e2b511261cd057cc0052141)

gottesmm commented:
@swift-ci test

gottesmm commented:
NOTE: Most of the changes in this PR are actually in the tests since I also massively expanded the amount of test coverage we have for this feature by changing most SendNonSendable tests to run both with and without NonisolatedNonsendingByDefault.

gottesmm commented:
So it is best to review the change commit by commit to see the actual amount of changes.

gottesmm committed:
[concurrency] Perform some fixes so that transfernonsendable_closureliterals_isolationinference.swift now passes.

The reason why this failed is that concurrently to @xedin landing
79af04ccc41faa717cfa3666e492f396993fcad1, I enabled
NonisolatedNonsendingByDefault on a bunch of other tests. That change broke the
test and so we needed to fix it.

This commit fixes a few issues that were exposed:

1. We do not propagate nonisolated(nonsending) into a closure if its inferred
context isolation is global actor isolated or if the closure captures an
isolated parameter. We previously just always inferred
nonisolated(nonsending). Unfortunately since we do not yet have capture
information in CSApply, this required us to put the isolation change into
TypeCheckConcurrency.cpp and basically have function conversions of the form:

```
(function_conversion_expr type="nonisolated(nonsending) () async -> Void"
   (closure_expr type="() async -> ()" isolated_to_caller_isolation))
```

Notice how we have a function conversion to nonisolated(nonsending) from a
closure expr that has an isolation that is isolated_to_caller.

2. With this in hand, we found that this pattern caused us to first thunk a
nonisolated(nonsending) function to an @concurrent function and then thunk that
back to nonisolated(nonsending), causing the final function to always be
concurrent. I put into SILGen a peephole that recognizes this pattern and emits
the correct code.

3. With that in hand, we found that we were emitting nonisolated(nonsending)
parameters for inheritActorContext functions. This was then fixed by @xedin in

With all this in hand, closure literal isolation and all of the other RBI tests
with nonisolated(nonsending) enabled pass.

rdar://154969621
(cherry picked from commit 648bb8fe3027a57e5ebd743c749744472f21b7b2)

gottesmm committed:
Update two tests for 4c1440735b8066d061c668d3b655955c1831f1dd.

I cherry-picked 4c1440735b8066d061c668d3b655955c1831f1dd to prevent merge
conflicts. I think these tests were updated by a subsequent PR. Rather than
cherry-picking more work, I just updated these tests since the change seemed
benign and that this kept the overall amount of cherry-picked commits small
(converting %$NUMBER to %kind$NUMBER for some ValueDecls).

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test windows platform

gottesmm commented:
windows failed due to:

gottesmm commented:
Windows failure was due to: https://github.com/swiftlang/swift-foundation/pull/1417

hborla reviewed (commented):
I'm still in the process of reviewing this change but I had a few comments in the first commit that I figured I'd post now.

xedin commented on test/Concurrency/attr_execution/migration_mode.swift:
```diff
@@ -393,3 +393,9 @@ do {
   }
 }
+
+// @_inheritActorContext prevents `nonisolated(nonsending)` inference.
```
What this comment is trying to say is that with `NonisolatedNonsendingByDefault` if parameter has `@_inheritActorContext` and it's type is `async` it won't get `nonisolated(nonsending)` inferred on it. `@_inheritActorContext` in such cases acts like `isolated` or `@isolated(any)` spelled on a parameter.

xedin replied:
I have removed changes to `@_inheritActorContext` from this PR.

gottesmm committed:
[rbi] Teach SendNonSendable how to more aggressively suppress sending errors around obfuscated Sendable functions

Specifically the type checker to work around interface types not having
isolation introduces casts into the AST that enrich the AST with isolation
information. Part of that information is Sendable. This means that we can
sometimes lose due to conversions that a function is actually Sendable. To work
around this, we today suppress those errors when they are emitted (post 6.2, we
should just change their classification as being Sendable... but I don't want to
make that change now).

This change just makes the pattern matching for these conversions handle more
cases so that transfernonsendable_closureliterals_isolationinference.swift now
passes.

xedin commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4234,6 +4234,11 @@ NeverNullType TypeResolver::resolveASTFunctionType(
 
+    if (options.contains(TypeResolutionFlags::InheritsActorContext)) {
+      diagnoseInvalid(repr, attr->getAttrLoc(),
+                      diag::inherit_actor_context_with_concurrent, attr);
```
I have removed changes to `@_inheritActorContext` from this PR.

xedin commented:
@swift-ci please test Linux platform
